contracts/SpritzPayV1.sol

```diff
@@ -66,38 +97,140 @@ contract SpritzPayV1 is SpritzPayStorage, Initializable, OwnableUpgradeable, Pau
+                });
+            }
+        }
+
+        address[] memory path = new address[](2);
+        path[0] = sourceTokenAddress;
+        path[1] = paymentTokenAddress;
+
+        uint256[] memory amounts;
+
+        //Execute the swap
+        if (!isNativeSwap) {
+            amounts = router.swapTokensForExactTokens(
+                paymentTokenAmount,
+                sourceTokenAmountMax,
+                path,
+                paymentRecipient,
+                block.timestamp
+            );
+        } else {
+            amounts = router.swapETHForExactTokens{ value: msg.value }(
+                paymentTokenAmount,
+                path,
+                paymentRecipient,
+                block.timestamp
+            );
+        }
+
+        uint256 sourceTokenSpentAmount = amounts[0];
+
+        require(amounts[1] == paymentTokenAmount, "Swap did not yield declared payment amount");
+
         emit Payment(
             paymentRecipient,
             _msgSender(),
             sourceTokenAddress,
-            sourceTokenAmount,
+            sourceTokenSpentAmount,
             paymentTokenAddress,
             paymentTokenAmount,
             paymentReference
         );
+
+        //Refund remaining balance left after the swap to the user
+        uint256 remainingBalance = sourceTokenAmountMax - sourceTokenSpentAmount;
+        if (remainingBalance > 0) {
+            bool refundSuccess = isNativeSwap
+                ? payable(_msgSender()).send(remainingBalance)
+                : sourceToken.safeTransfer(_msgSender(), remainingBalance);
+
+            if (!refundSuccess) {
+                revert FailedRefund({ tokenAddress: sourceTokenAddress, amount: remainingBalance });
+            }
+        }
+    }
+
+    /**
+     * @notice Authorizes the swap router to spend a token
+     * @param token Address of an ERC20 token
+     * @param allowanceTarget Target contract which can spend this contract's token balance
+     */
+    function approveTokenSpend(address token, address allowanceTarget) private returns (bool approveSuccess) {
+        IERC20 erc20 = IERC20(token);
+        uint256 max = 2**256 - 1;
+        return erc20.safeApprove(allowanceTarget, max);
+    }
+
+    /**
+     * @notice Transfers token to
+     * @param token Address of an ERC20 used for payment
+     * @param from Address from which to withdraw
+     * @param to Address which receives token
+     * @param amount Amount to withdraw
+     */
+    function safeTransferToken(
+        address token,
+        address from,
+        address to,
+        uint256 amount
+    ) internal returns (bool transferSuccess) {
+        IERC20 erc20 = IERC20(token);
+        return erc20.safeTransferFrom(from, to, amount);
     }
 
     /*
-     * Admin functions to edit the admin, router address or weth address
+     * Admin functions
      */
 
+    receive() external payable {}
+
+    fallback() external payable {}
```

Comment:
Interesting. Do we need both? receive should be enough if we are just looking to get back leftover funds from uni